Use nullptr for main source filename in clang_parseTranslationUnit2FullArgv and rely on -working-directory for relative filename resolution

MaskRay opened the pull request:
… that we do not need to strip main source filename from args

jacobdufault commented:
This still causes parsing failures for chrome.

```
2017-11-29 11:09:15.473 (  34.610s) [indexer34       ]clang_translation_unit.cc:37     ERR| libclang generic failure for /work2/chrome/src/services/network/public/cpp/net_adapters.cc with args ../../third_party/llvm-build/Release+Asserts/bin/clang++, -xc++, -DV8_DEPRECATION_WARNINGS, -DDCHECK_ALWAYS_ON=1, -DUSE_UDEV, -DUSE_AURA=1, -DUSE_NSS_CERTS=1, -DUSE_OZONE=1, -DFULL_SAFE_BROWSING, -DSAFE_BROWSING_CSD, -DSAFE_BROWSING_DB_LOCAL, -DCHROMIUM_BUILD, -DFIELDTRIAL_TESTING_ENABLED, -D_FILE_OFFSET_BITS=64, -D_LARGEFILE_SOURCE, -D_LARGEFILE64_SOURCE, -DCR_CLANG_REVISION="318369-1", -D__STDC_CONSTANT_MACROS, -D__STDC_FORMAT_MACROS, -DCOMPONENT_BUILD, -DOS_CHROMEOS, -DNDEBUG, -DNVALGRIND, -DDYNAMIC_ANNOTATIONS_ENABLED=0, -DU_USING_ICU_NAMESPACE=0, -DU_ENABLE_DYLOAD=0, -DICU_UTIL_DATA_IMPL=ICU_UTIL_DATA_FILE, -DUCHAR_TYPE=uint16_t, -DGOOGLE_PROTOBUF_NO_RTTI, -DGOOGLE_PROTOBUF_NO_STATIC_INITIALIZER, -DHAVE_PTHREAD, -DPROTOBUF_USE_DLLS, -DBORINGSSL_SHARED_LIBRARY, -I/work2/chrome/src, -I/work2/chrome/src/out/Release/gen, -I/work2/chrome/src/third_party/ced/src, -I/work2/chrome/src/third_party/icu/source/common, -I/work2/chrome/src/third_party/icu/source/i18n, -I/work2/chrome/src/third_party/protobuf/src, -I/work2/chrome/src/third_party/protobuf/src, -I/work2/chrome/src/out/Release/gen/protoc_out, -I/work2/chrome/src/third_party/boringssl/src/include, -I/work2/chrome/src/build/linux/debian_stretch_amd64-sysroot/usr/include/nss, -I/work2/chrome/src/build/linux/debian_stretch_amd64-sysroot/usr/include/nspr, -fno-strict-aliasing, -Wno-builtin-macro-redefined, -D__DATE__=, -D__TIME__=, -D__TIMESTAMP__=, -funwind-tables, -fPIC, -pipe, -B../../third_party/binutils/Linux_x64/Release/bin, -pthread, -fcolor-diagnostics, -no-canonical-prefixes, -m64, -march=x86-64, -Wall, -Werror, -Wextra, -Wno-missing-field-initializers, -Wno-unused-parameter, -Wno-c++11-narrowing, -Wno-covered-switch-default, -Wno-unneeded-internal-declaration, -Wno-inconsistent-missing-override, -Wno-undefined-var-template, -Wno-nonportable-include-path, -Wno-address-of-packed-member, -Wno-unused-lambda-capture, -Wno-user-defined-warnings, -Wno-enum-compare-switch, -Wno-tautological-unsigned-zero-compare, -Wno-null-pointer-arithmetic, -Wno-tautological-constant-compare, -Wtautological-constant-out-of-range-compare, -O2, -fno-ident, -fdata-sections, -ffunction-sections, -fno-omit-frame-pointer, -g0, -fvisibility=hidden, -Wheader-hygiene, -Wstring-conversion, -Wtautological-overlap-compare, -Wno-header-guard, -std=gnu++14, -fno-rtti, -nostdinc++, -isystem/work2/chrome/src/buildtools/third_party/libc++/trunk/include, -isystem/work2/chrome/src/buildtools/third_party/libc++abi/trunk/include, --sysroot=/work2/chrome/src/build/linux/debian_stretch_amd64-sysroot, -fno-exceptions, -fvisibility-inlines-hidden, ../../services/network/public/cpp/net_adapters.cc, -resource-dir=/usr/local/google/home/jdufault/cquery-wip/clang_resource_dir
```

MaskRay commented:
Is there more diagnostic?

jacobdufault commented:
No, unfortunately. clang doesn't emit any errors or explain why it failed.

MaskRay commented:
Rebased. Does your Chrome build give more diagnostics now?

jacobdufault commented on src/project.cc:
```diff
@@ -825,7 +828,7 @@ TEST_SUITE("Project") {
          "-fno-exceptions",
          "-fvisibility-inlines-hidden",
-         "-xc++",
+         "../../ash/login/ui/lock_screen_sanity_unittest.cc",
```
This path should probably be normalized (ie, it should have & before it).

MaskRay replied:
What is the `&` for?

jacobdufault replied:
It means that the path was normalized. Because it is a test normalization doesn't actually happen, just & is prepended.

jacobdufault replied:
ie, the path should be `"&../../ash/login/ui/lock_screen_sanity_unittest.cc"`

MaskRay replied:
Does clang USR encode absolute filename? If not, is it fine to leave filename arguments as is?

MaskRay replied:
The main source filename is returned by `clang_CompileCommand_getFilename(cx_command)`. It is difficult to recognize the main source filename (thinking of `-include -isystem -I`) and put a `&` before it to mark it as normalized (for testing purpose).

jacobdufault replied:
This still fails because the path is not being normalized.  The filename should be in `entry.file`.

One alternate approach that does not depend on filename (though I'd prefer checking the filename):
```
if (arg[0] != '-')
  arg = cleanup_maybe_relative_path(arg);
```

MaskRay replied:
Rebased.  Added these two lines.

MaskRay replied:
After switching to `-working-directory`, this normalization is removed.

MaskRay commented:
Rebased.

jacobdufault commented on src/project.cc:
```diff
@@ -138,12 +151,11 @@ Project::Entry GetCompilationEntryFromCompileCommandEntry(
+      std::string arg1 = cleanup_maybe_relative_path(arg);
       if (add_next_flag_to_quote_dirs)
-        config->quote_dirs.insert(arg);
+        config->quote_dirs.insert(arg1);
```
Instead of `arg1` (which doesn't convey any extra meaning) use something a bit more descriptive like `normalized_path` or similar.

jacobdufault commented on src/project.cc:
```diff
@@ -290,12 +296,8 @@ std::vector<Project::Entry> LoadCompilationEntriesFromDirectory(
 
     our_time.Resume();
     std::string absolute_filename;
```
Is `absolute_filename` still used? It doesn't seem to be initialized anymore. Remove it?

jacobdufault commented:
Still broken on chrome, I'm not sure why yet though. I'll have to spend some time investigating since I believe this approach should work.

MaskRay commented:
Rebased. Around project.cc:109, changed

``result.filename = entry.file;`
to
`result.filename = cleanup_maybe_relative_path(entry.file);`

Otherwise the cache file may be incorrect.

MaskRay commented:
Does this PR work on Chrome?

jacobdufault commented:
Still fails.

I wonder if we should normalize the path to the clang binary?

Example failure when running the arguments via command line:

```
/work2/chrome/src/third_party/llvm-build/Release+Asserts/bin/clang++ -working-directory /work2/chrome/src/out/Release -xc++ -DV8_DEPRECATION_WARNINGS -DDCHECK_ALWAYS_ON=1 -DUSE_UDEV -DUSE_AURA=1 -DUSE_NSS_CERTS=1 -DUSE_OZONE=1 -DFULL_SAFE_BROWSING -DSAFE_BROWSING_CSD -DSAFE_BROWSING_DB_LOCAL -DCHROMIUM_BUILD -DFIELDTRIAL_TESTING_ENABLED -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -D_LARGEFILE64_SOURCE -DCR_CLANG_REVISION="318667-1" -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -DCOMPONENT_BUILD -DOS_CHROMEOS -DNDEBUG -DNVALGRIND -DDYNAMIC_ANNOTATIONS_ENABLED=0 -DGOOGLE_PROTOBUF_NO_RTTI -DGOOGLE_PROTOBUF_NO_STATIC_INITIALIZER -DHAVE_PTHREAD -DPROTOBUF_USE_DLLS -I../.. -Igen -I../../third_party/protobuf/src -Igen/protoc_out -I../../third_party/protobuf/src -fno-strict-aliasing -Wno-builtin-macro-redefined -D__DATE__= -D__TIME__= -D__TIMESTAMP__= -funwind-tables -fPIC -pipe -B../../third_party/binutils/Linux_x64/Release/bin -pthread -fcolor-diagnostics -no-canonical-prefixes -m64 -march=x86-64 -Wall -Werror -Wextra -Wno-missing-field-initializers -Wno-unused-parameter -Wno-c++11-narrowing -Wno-covered-switch-default -Wno-unneeded-internal-declaration -Wno-inconsistent-missing-override -Wno-undefined-var-template -Wno-nonportable-include-path -Wno-address-of-packed-member -Wno-unused-lambda-capture -Wno-user-defined-warnings -Wno-enum-compare-switch -Wno-tautological-unsigned-zero-compare -Wno-null-pointer-arithmetic -Wno-tautological-constant-compare -Wtautological-constant-out-of-range-compare -O2 -fno-ident -fdata-sections -ffunction-sections -fno-omit-frame-pointer -g0 -fvisibility=hidden -Wheader-hygiene -Wstring-conversion -Wtautological-overlap-compare -std=gnu++14 -fno-exceptions -fno-rtti -nostdinc++ -isystem../../buildtools/third_party/libc++/trunk/include -isystem../../buildtools/third_party/libc++abi/trunk/include --sysroot=../../build/linux/debian_stretch_amd64-sysroot -fvisibility-inlines-hidden gen/json.pb.cc -resource-dir=/usr/local/google/home/jdufault/cquery/build/release/lib/clang+llvm-5.0.1-x86_64-linux-gnu-ubuntu-14.04/lib/clang/5.0.1 -Wno-unknown-warning-option -fparse-all-comments -fspell-checking
```

==>

```
In file included from /work2/chrome/src/out/Release/gen/json.pb.cc:5:
In file included from /work2/chrome/src/out/Release/gen/json.pb.h:7:
In file included from /work2/chrome/src/out/Release/../../buildtools/third_party/libc++/trunk/include/string:469:
In file included from /work2/chrome/src/out/Release/../../buildtools/third_party/libc++/trunk/include/__config:179:
/work2/chrome/src/out/Release/../../build/linux/debian_stretch_amd64-sysroot/usr/include/features.h:364:12: fatal error: 'sys/cdefs.h' file not found
#  include <sys/cdefs.h>
           ^~~~~~~~~~~~~
1 error generated.
```

MaskRay commented:
It it a cross compiler? Do you where `sys/cdefs.h` resides and how that path is relative to the clang driver?

jacobdufault commented on src/project.cc:
```diff
@@ -307,13 +312,8 @@ std::vector<Project::Entry> LoadCompilationEntriesFromDirectory(
     // LOG_S(INFO) << "Got args " << StringJoin(entry.args);
 
     our_time.Resume();
-    std::string absolute_filename;
```
Why did you change entry.file to be relative? I think this should stay absolute, so ie, reading the file before passing to clang work as expected. If I revert this part of the change it seems to work fine on chrome.

jacobdufault replied:
I suspect this patch assumes that cquery working dir is the same as the project working dir (which is not always true, ie, in the case of vscode).

MaskRay replied:
Reverted this change.

jacobdufault commented:
:)
